### ISSUE TITLE
dart vm 'unreachable code' exception in Scavenge

### ISSUE THREAD

**Issue by aam** [dart-lang/sdk#34301]
Flutter buildbot failed on https://cirrus-ci.com/task/4728162248294400 with dart vm exception:

```
03:30 +199: test/base/logger_test.dart: Spinners AnsiStatus works when cancelled                                                                                                                       Class Id: 0
../../third_party/dart/runtime/vm/raw_object.cc: 280: error: unreachable code
Dumping native stack trace for thread 1c2f
  [0x000000010b3357a5] dart::Profiler::DumpStackTrace(void*)
  [0x000000010b3357a5] dart::Profiler::DumpStackTrace(void*)
  [0x000000010b5a7d42] dart::Assert::Fail(char const*, ...)
  [0x000000010b348dd1] dart::RawObject::VisitPointersPredefined(dart::ObjectPointerVisitor*, long)
  [0x000000010b566ed4] dart::Scavenger::ProcessToSpace(dart::ScavengerVisitor*)
  [0x000000010b5677cc] dart::Scavenger::Scavenge()
  [0x000000010b55c179] dart::Heap::CollectNewSpaceGarbage(dart::Thread*, dart::Heap::GCReason)
  [0x000000010b55b005] dart::Heap::AllocateNew(long)
  [0x000000010b23f2e2] dart::Object::Allocate(long, long, dart::Heap::Space)
  [0x000000010b2a435c] dart::Instance::New(dart::Class const&, dart::Heap::Space)
  [0x000000010b38c22d] dart::DRT_AllocateObject(dart::NativeArguments)
  [0x000000010e78091b] Unknown symbol
  [0x0000000119a08329] Unknown symbol
  [0x0000000137106455] Unknown symbol
  [0x0000000137117403] Unknown symbol
  [0x0000000137102e43] Unknown symbol
  [0x00000001371182b4] Unknown symbol
  [0x0000000134bfbc01] Unknown symbol
  [0x00000001371176f8] Unknown symbol
  [0x0000000134bdce78] Unknown symbol
  [0x00000001371030dc] Unknown symbol
  [0x0000000137117237] Unknown symbol
  [0x0000000137107c02] Unknown symbol
  [0x000000013711689e] Unknown symbol
  [0x0000000137100fe8] Unknown symbol
  [0x000000013147da45] Unknown symbol
  [0x0000000134bdc794] Unknown symbol
  [0x0000000134bdce78] Unknown symbol
  [0x00000001371030dc] Unknown symbol
  [0x0000000137117237] Unknown symbol
  [0x0000000137107c02] Unknown symbol
  [0x000000013711689e] Unknown symbol
  [0x0000000137100fe8] Unknown symbol
  [0x000000013710597c] Unknown symbol
  [0x0000000132646560] Unknown symbol
  [0x0000000137101d18] Unknown symbol
  [0x00000001371070d1] Unknown symbol
  [0x0000000134bfcaa3] Unknown symbol
  [0x00000001371030dc] Unknown symbol
  [0x000000013710a517] Unknown symbol
  [0x0000000137106074] Unknown symbol
  [0x0000000137116b1b] Unknown symbol
  [0x0000000137116eab] Unknown symbol
  [0x00000001326438b8] Unknown symbol
  [0x00000001371030dc] Unknown symbol
  [0x000000013710a517] Unknown symbol
  [0x0000000137105f9a] Unknown symbol
  [0x0000000137116b1b] Unknown symbol
  [0x0000000137101645] Unknown symbol
  [0x0000000134bd3e26] Unknown symbol
  [0x0000000134bd3b96] Unknown symbol
  [0x0000000137100682] Unknown symbol
  [0x00000001371045af] Unknown symbol
  [0x0000000134bd0a7b] Unknown symbol
  [0x0000000137105492] Unknown symbol
  [0x000000010e780da3] Unknown symbol
  [0x000000010b1dc8b0] dart::DartEntry::InvokeFunction(dart::Function const&, dart::Array const&, dart::Array const&, unsigned long)
  [0x000000010b1dfd9f] dart::DartLibraryCalls::HandleMessage(dart::Object const&, dart::Instance const&)
  [0x000000010b206952] dart::IsolateMessageHandler::HandleMessage(dart::Message*)
  [0x000000010b23798a] dart::MessageHandler::HandleMessages(dart::MonitorLocker*, bool, bool)
  [0x000000010b23815c] dart::MessageHandler::TaskCallback()
  [0x000000010b3e145d] dart::ThreadPool::Worker::Loop()
  [0x000000010b3e128f] dart::ThreadPool::Worker::Main(unsigned long)
  [0x000000010b2e4829] dart::ThreadStart(void*)
  [0x00007fff6477b661] _pthread_body
  [0x00007fff6477b50d] _pthread_body
  [0x00007fff6477abf9] thread_start
-- End of DumpStackTrace
🕐 ELAPSED TIME: 0:03:33.062542 for ../../bin/cache/dart-sdk/bin/pub run test -rcompact -j1 --no-color in packages/flutter_tools: 
```

cc @rmacnak-google

**Comment by rmacnak-google** [comment]
I left this running in a loop in gdb overnight but was unable to reproduce.

**Comment by dnfield** [comment]
This has been happening again in TAP - https://sponge.corp.google.com/target?id=940ac991-a028-441b-8b98-ed5355819608&target=//third_party/dart/flutter_tools:tests_base_logger_test and apparently it's reproducing pretty consistently with that particular changeset.

**Comment by mraleph** [comment]
@rmacnak-google could you take a look given that @dnfield and @willlarche are able to reliably reproduce it?

**Comment by mraleph** [comment]
I am also marking this as P0 because it is blocking the internal Flutter roll.

**Comment by willlarche** [comment]
I am able to reproduce it ~100% on TAP and 98-99% on local blaze test.

**Comment by willlarche** [comment]
Disabling test internally.

**Comment by rmacnak-google** [comment]
While visiting the spill slots of a frame with optimized code for dart:async.Future.doWhile.<anonymous closure>, we encounter either a non-object or a corrupt object.

**Comment by rmacnak-google** [comment]
The stack map tells us to visit rbp-0x80, but this slot is never referenced in the generated code.
https://gist.github.com/rmacnak-google/0fdf03e7a0682954b3b02a2c34b08051#file-gistfile1-txt-L9

**Comment by mraleph** [comment]
I stared at the code that Ryan provided and I figured out that this is exactly the same bug that caused us to disable multiple entry points. What confused me here is that internally the version of Dart used to run tests is different from the version of Dart in the Flutter engine. The version of Dart in the engine actually has this issue fixed - but this has not been rolled, so internal version of Dart SDK actually still has the issue.
